Add troubleshooting guide (#2563)
microsoft/ebpf-for-windows: docs/TroubleshootingGuide.md

2. [Verify WFP objects are present](#verify-wfp-objects-are-present)
3. [Verify the eBPF Program is Configured Correctly](#verify-the-ebpf-program-is-configured-correctly)

--------------------

## Verify eBPF Components Are Running

Verify that the necessary services are running. Run the following commands:
```
sc.exe queryex netebpfext
sc.exe queryex ebpfcore
```
We expect to see the following output, notably that the service is in the **Running** state:
```
SERVICE_NAME: ebpfcore
        TYPE               : 1  KERNEL_DRIVER
        STATE              : 4  RUNNING
                                (STOPPABLE, NOT_PAUSABLE, IGNORES_SHUTDOWN)
        WIN32_EXIT_CODE    : 0  (0x0)
        SERVICE_EXIT_CODE  : 0  (0x0)
        CHECKPOINT         : 0x0
        WAIT_HINT          : 0x0
        PID                : 0
        FLAGS              :
```

**Mitigation:** For each service that is not running, execute:
```
sc.exe start netebpfext
sc.exe start ebpfcore
```

If the problem persists, obtain the `SERVICE_EXIT_CODE` and look at the [eBPF diagnostic traces](./Diagnostics.md#ebpf-diagnostic-traces) for further diagnosis.

--------------------

## Verify WFP objects are present

> netebpfext uses the WFP platform to implement certain eBPF program types. If you are observing issues with the eBPF program not getting invoked at all, you should check if the necessary WFP objects are present.

Reviewer: ```suggestion
netebpfext.sys uses the WFP platform to implement certain eBPF program types. If you are observing issues with the eBPF program not getting invoked at all, you should check if the necessary WFP objects are present.
```

4. [eBPF Program Load or Attach Failures](#ebpf-program-load-or-attach-failures)

--------------------

### Verify the eBPF Program is Loaded

To check that the eBPF program is loaded, execute:
```
bpftool.exe -p prog
```
In this output, check that you see the expected eBPF program, looking at the `name` and `type`. Take note of the `id` and `map_ids` for the next set of checks.

Example Output:
```
[{
    "id": 196867,
    "type": "sock_addr",
    "name": "authorize_connect4",
    "map_ids": [66054,131331]
}]
```

--------------------

### Verify the eBPF Program is Attached

To check that the eBPF program is attached, execute:
```
bpftool.exe -p link
```
In this output, check for an entry with the `prog_id` which matches the `id` from the above output, and confirm that the `attach_type` is as expected.

Example output:
```
bpftool.exe -p link
[{
    "id": 262403,
    "type": 2,
    "prog_id": 196867,
    "cgroup_id": 0,
    "attach_type": "cgroup/connect4"
}]
```

--------------------

### Verify eBPF Maps are Properly Configured

To check the map content, execute:
```
bpftool.exe -p map show id <id>
```

> In this output, use the `map_ids` from the above output. Map usage is up to the eBPF developer, so you should confirm that the `type` and `name` is as expected for the scenario. This example output is from invoking the bpftool for each map:

Reviewer: ```suggestion
In this output, use the `map_ids` from the above output. Map usage is up to the eBPF program developer, so you should confirm that the `type` and `name` is as expected for the scenario. This example output is from invoking the bpftool for each map:
```

```
{
    "id": 66054,
    "type": "hash",
    "name": "policy_map",
    "flags": 0,
    "bytes_key": 24,
    "bytes_value": 24,
    "max_entries": 10
}

{
    "id" : 131331,
    "type" : "lru_hash",
    "name" : "audit_map",
    "flags" : 0,
    "bytes_key" : 8,
    "bytes_value" : 24,
    "max_entries" : 1000
}
```

Once you have confirmed that the expected maps are present, you can then dump the map entries and check that the values are as expected. You will need the `map_ids` from above. Then, you can execute the following command:
```
bpftool.exe map dump id <id>
```

Example Output:
```
key:
08 08 08 08 00 00 00 00  00 00 00 00 00 00 00 00
1a 0a 00 00 06 00 00 00
value:
7f 00 00 01 00 00 00 00  00 00 00 00 00 00 00 00
15 b3 00 00 00 00 00 00
Found 1 element
```

The map usage is up to the eBPF program developer. You should follow up with the developer to understand what structures are used in the map and how you can use this output to verify that the map entries are populated correctly.

--------------------

### eBPF Program Load or Attach Failures

> Once you have [identified that the program is either not attached or not loaded](#troubleshooting-general-ebpf-program-issues), you should first confirm that the eBPF client has attempted to load and attach the program (i.e there were no issues within the eBPF client itself). If you have confirmed that the eBPF client has attempted to load/attach the program, but it has failed, you can use the following to further debug your issue.

Reviewer: ```suggestion
Once you have [identified that the program is either not attached or not loaded](#troubleshooting-general-ebpf-program-issues), you should first confirm that the eBPF client has attempted to load and attach the program (i.e., there were no issues within the eBPF client itself). If you have confirmed that the eBPF client has attempted to load/attach the program, but it has failed, you can use the following to further debug your issue.
```
also add source line breaks to make github diffs easier in the future

The common flow for configuring a eBPF program would be to first `open` the program, then `load` the program, and finally, `attach` the program. For each of these operations, you can look for a trace statement within the [eBPF diagnostic traces](./Diagnostics.md#ebpf-diagnostic-traces) which indicates failure:
- Open: Look for a trace with `ebpf_object_open`
- Load: Look for a trace with `ebpf_object_load`
- Attach: Look for a trace with `ebpf_program_attach_by_fd`

There are a few classes of known issues:

**eBPF Client Issues**

There are certain errors that likely point to the eBPF client. These errors will be present in [eBPF diagnostic traces](./Diagnostics.md#ebpf-diagnostic-traces):

> - `ERROR_ACCESS_DENIED` or `STATUS_ACCESS_DENIED`. This means that the eBPF client is not running as admin or localsystem. This points to an issue with the eBPF client. The resolution here is to run your eBPF program or service as localsystem or admin.

Reviewer: ```suggestion
- `ERROR_ACCESS_DENIED` or `STATUS_ACCESS_DENIED`. This means that the user-mode application is not running as admin or localsystem. This points to an issue with the eBPF client. The resolution here is to run the user-mode application or service as localsystem or admin.
```

> - `ERROR_FILE_NOT_FOUND`. This indicates that the eBPF client tried to register a program with an invalid path. This points to an issue within the eBPF client. The resolution is to change the path used by the eBPF client.

Reviewer: ```suggestion
- `ERROR_FILE_NOT_FOUND`. This indicates that the application tried to open an eBPF program with an invalid path. This points to an issue within the application. The resolution is to change the path used by the application.
```

2. This trace is present and has the expected IP properties, but traffic is still not reaching the proxy. Please [check for interoperability issues with another WFP callout](#interoperability-issues-with-another-wfp-callout).
3. This trace is not present at all. First, check the [eBPF diagnostic traces](./Diagnostics.md#ebpf-diagnostic-traces) to identify if there were any issues within the callout itself. If there are no errors in this codepath, [check for interoperability issues with another WFP callout](#interoperability-issues-with-another-wfp-callout).

--------------------

#### Interoperability Issues With Another WFP Callout

Multiple WFP callouts at the connect redirect layer may cause unexpected results. This may surface as one of the following symptoms:
1. The connection is not reaching the proxy. This can happen both even when the eBPF callout is getting invoked, but also when it does not get invoked.
2. The connection reaches the proxy, but does not reach the expected final destination.
3. Kernel crash

> To check if there is another WFP callout at the connect redirect layer, you should search in the [WFP state diagnostic file](./Diagnostics.md#wfp-state) for the string `FWPM_LAYER_ALE_CONNECT_REDIRECT_V4` (or `V6`, if applicable). Within this layer, you can look in the `callouts` section of the file. We expect to see only 1 eBPF callout here. If you see more than 1, then another WFP callout driver may be attempting to redirect the same connections that you eBPF program is, which may affect the final connection.

Reviewer: ```suggestion
To check if there is another WFP callout at the connect redirect layer, you should search in the [WFP state diagnostic file](./Diagnostics.md#wfp-state) for the string `FWPM_LAYER_ALE_CONNECT_REDIRECT_V4` (or `V6`, if applicable). Within this layer, you can look in the `callouts` section of the file. We expect to see only 1 eBPF callout here. If you see more than 1, then another WFP callout driver may be attempting to redirect the same connections that your eBPF program is, which may affect the final connection.
```
also break into multiple source lines

This document contains a troubleshooting guide for issues related to eBPF.

--------------------

# What Kind of Issue Are You Having ?

> - [The eBPF program is not getting invoked](#troubleshooting-general-ebpf-program-issues)
> - [A specific eBPF program is not behaving as expected](#troubleshooting-issues-related-to-a-specific-program-type)
> - [A specific eBPF program is failing verification](./debugging.md)

Reviewer: ```suggestion
- [A specific eBPF program is failing verification](./debugging.md)
- [The eBPF program is not getting invoked](#troubleshooting-general-ebpf-program-issues)
- [A specific eBPF program is not behaving as expected](#troubleshooting-issues-related-to-a-specific-program-type)
```
list in logically chronological order

    <displayData>
        <name>net eBPF sock_addr hook</name>
        <description>net eBPF sock_addr hook WFP filter</description>
    </displayData>
    <flags/>
    <providerKey>{ddb851f5-841a-4b77-8a46-bb7063e9f162}</providerKey>
    <providerData/>
    <layerKey>FWPM_LAYER_ALE_AUTH_CONNECT_V4</layerKey>
    <subLayerKey>{7c7b3fb9-3331-436a-98e1-b901df457fff}</subLayerKey>
    <weight>
        <type>FWP_EMPTY</type>
    </weight>
    <filterCondition/>
    <action>
        <type>FWP_ACTION_CALLOUT_TERMINATING</type>
        <filterType>{98849e0b-b07d-11ec-9a30-18602489beee}</filterType>
    </action>
    <rawContext>18446624845314639248</rawContext>
    <reserved/>
    <filterId>68244</filterId>
    <effectiveWeight>
    <type>FWP_UINT64</type>
    <uint64>0</uint64>
    </effectiveWeight>
</item>
```

--------------------

## Verify the eBPF Program is Configured Correctly

1. [Verify the eBPF program is loaded](#verify-the-ebpf-program-is-loaded)
2. [Verify the eBPF program is attached](#verify-the-ebpf-program-is-attached)
3. [Verify eBPF maps are properly configured](#verify-ebpf-maps-are-properly-configured)

> 4. [eBPF Program Load or Attach Failures](#ebpf-program-load-or-attach-failures)

Reviewer: All the other bullets are written as an instruction to the user.  "eBPF Program Load or Attach Failures" is not an instruction to the user.

> 1. [Verify the eBPF program is loaded](#verify-the-ebpf-program-is-loaded)
> 2. [Verify the eBPF program is attached](#verify-the-ebpf-program-is-attached)
> 3. [Verify eBPF maps are properly configured](#verify-ebpf-maps-are-properly-configured)
> 4. [eBPF Program Load or Attach Failures](#ebpf-program-load-or-attach-failures)
> 5. [Verify the eBPF program passes the verifier](./debugging.md)

Reviewer: ```suggestion
1. [Verify the eBPF program passes the verifier](./debugging.md)
2. [Verify the eBPF program is loaded](#verify-the-ebpf-program-is-loaded)
3. [Verify the eBPF program is attached](#verify-the-ebpf-program-is-attached)
4. [Verify eBPF maps are properly configured](#verify-ebpf-maps-are-properly-configured)
5. [eBPF Program Load or Attach Failures](#ebpf-program-load-or-attach-failures)
```
List in chronological order in the lifetime of a program
